src/lsp.ts

```diff
@@ -0,0 +1,180 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  private config: WorkspaceConfiguration;
+  private ls: LanguageClient | null = null;
+  private executableName: string;
+  private context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+  abstract repoURL: string;
+  protected abstract get debugExe(): Executable;
+  protected abstract get runExe(): Executable;
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+
+    if (this.languageServerPath == null)
+      return;
+
+    this.startLanguageServer();
+  }
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+
+    if (this.languageServerPath == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  private startLanguageServer() {
+    const serverOptions: ServerOptions = {
+      run: this.runExe,
+      debug: this.debugExe,
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient(this.executableName, `Meson Language Server (${this.executableName})`, serverOptions, this.clientOptions, true);
+    this.ls.start();
+  }
+
+  canDownloadLanguageServer(): boolean {
+    return this.downloadInfo != null && this.findLanguageServer() == null
+  }
+
+  get languageServerPath(): string | null {
+    return this.config.languageServerPath || this.findLanguageServer() || which.sync(this.executableName, { nothrow: true })
+  }
+
+  async setupLanguageServer(): Promise<void> {
+    const lspDir = path.join(this.getExtensionLSPDir(), this.executableName);
+    const downloadInfo = this.downloadInfo;
+    fs.rmSync(
+      lspDir,
+      {
+        recursive: true,
+        force: true
+      }
+    );
+    fs.mkdirSync(
+      lspDir,
+      { recursive: true }
+    );
+    const tmpPath = path.join(os.tmpdir(), `lsp-${Date.now()}.zip`);
+    try {
+      this.downloadFile(downloadInfo[0], tmpPath).then(() => {
+        this.computeFileHash(tmpPath).then(str => {
+          const expected = downloadInfo[1];
```

Comment:
I think when you have named tuples, you can use `downloadInfo.hash`

Reply:
No, does not seem to work:

```

> mesonbuild@1.9.0 compile
> tsc -p ./

src/lsp/index.ts:84:41 - error TS2339: Property 'hash' does not exist on type '[url: string, hash: string]'.

84           const expected = downloadInfo.hash;
                                           ~~~~


Found 1 error in src/lsp/index.ts:84
```

Reply:
Interesting. Had no idea. I guess it is purely for documentation purposes then.